libflux/flux-core/src/semantic/fresh.rs

```diff
@@ -97,38 +149,54 @@ impl Fresh for Tvar {
     fn fresh(self, f: &mut Fresher, sub: &mut TvarMap) -> Self {
         *sub.entry(self).or_insert_with(|| f.fresh())
     }
+    fn fresh_ref(&self, f: &mut Fresher, sub: &mut TvarMap) -> Option<Self> {
+        Some(*sub.entry(*self).or_insert_with(|| f.fresh()))
+    }
 }
 
 impl Fresh for Array {
     fn fresh(self, f: &mut Fresher, sub: &mut TvarMap) -> Self {
         Array(self.0.fresh(f, sub))
     }
+    fn fresh_ref(&self, f: &mut Fresher, sub: &mut TvarMap) -> Option<Self> {
+        self.0.fresh_ref(f, sub).map(Array)
+    }
 }
 
 impl Fresh for Record {
-    fn fresh(mut self, f: &mut Fresher, sub: &mut TvarMap) -> Self {
+    fn fresh(self, f: &mut Fresher, sub: &mut TvarMap) -> Self {
+        self.fresh_ref(f, sub).unwrap_or(self)
+    }
+    fn fresh_ref(&self, f: &mut Fresher, sub: &mut TvarMap) -> Option<Self> {
         let mut props = MonoTypeVecMap::new();
         let mut extends = false;
         let mut tv = Tvar(0);
+        let mut cur = self;
         loop {
-            match self {
+            match cur {
                 Record::Empty => {
                     break;
                 }
                 Record::Extension {
                     head,
                     tail: MonoType::Record(b),
                 } => {
-                    props.entry(head.k).or_insert_with(Vec::new).push(head.v);
-                    self = *b;
+                    props
+                        .entry(head.k.clone())
+                        .or_insert_with(Vec::new)
+                        .push(head.v.clone());
+                    cur = b;
```

Comment:
Is this formatting automatic with a format command? Looks nice.

Reply:
Yes, I always have format on save on so the code (should) always be formatted.